### PR TITLE
Convert to TypeScript

### PR DESCRIPTION
**Pull request by juhanakristian**
**What**:

Converting library to TypeScript, issue #498
Removes deprecated wait util (as per discussion in KCD Discord)
https://discord.com/channels/715220730605731931/785649901782433852/786338242101379142

Creating this PR on behalf of @tigerabrodi we and other people from KCD Discord are going to work on it as a group

**Checklist**:

- [x] Documentation updated
- [x] Tests
- [x] Ready to be merged
      
- [x] Added myself to contributors table

### CODE REVIEW

**Comment by juhanakristian** [comment]
@mpeyper The issues you listed in the previous PR have been resolved and if there isn't anything else this should be ready for merging. Let us know what you think :)

**Comment by mpeyper** [comment]
Thanks, I'll take a look today.

File: src/pure.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  } catch (err) {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+    if (err.then) {
+      throw err
+    } else {
+      onError(err)
+    }
+  }
+  return null
+}
```
Comment (mpeyper) [inline src/pure.tsx:35]: The catch block was bothering me so I tried something... I _think_ it's better (it's definitely more explicit), but would like some feedback on it:

```suggestion
 
function isPromise<T>(value: unknown) : value is PromiseLike<T> {
  return typeof (value as PromiseLike<T>).then === 'function'
}

function TestHook<TProps, TResult>({
  callback,
  hookProps,
  onError,
  children
}: TestHookProps<TProps, TResult>) {
  try {
    if (hookProps) {
      children(callback(hookProps))
    } else {
      // coerce into undefined, so it maintains the previous behaviour
      children(callback((undefined as unknown) as TProps))
    }
  } catch (err: unknown) {
    if (isPromise(err)) {
      throw err as unknown
    } else {
      onError(err as Error)
    }
  }
  return null
}
```

(I'm not sure if you can do suggestions for multiple seperate blocks, so just focus on the `catch` block and `isPromise` function).

**Review by mpeyper — COMMENTED** [review]
This is looking very close.  A couple of comments here, mostly minor stuff.

File: src/pure.tsx [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  onError,
+  children
+}: TestHookProps<TProps, TResult>) {
+  try {
+    if (hookProps) {
+      children(callback(hookProps))
+    } else {
+      // coerce into undefined, so it maintains the previous behaviour
+      children(callback((undefined as unknown) as TProps))
+    }
```
[the remark this answers is not in the document]
Comment (nobrayner) [reply src/pure.tsx:24]: @mpeyper 🤦‍♂️ Thanks... I guess I got stuck down a rabit hole for that 😛

File: src/pure.tsx (hunk 1 above)
Comment (nobrayner) [reply src/pure.tsx:35]: Why is the promise error being cast back to unknown, after you went to the effort of Type Guarding it? You may as well just not Type Guard it and just keep err as the unknown type.

Comment (mpeyper) [reply src/pure.tsx]: Yeah, it was _conceptually_ more correct than being more useful.  Perhaps leaving the return type as just `boolean` on `isPromise` is a bit easier.

**Comment by mpeyper** [comment]
ok, it's been long enough... I'm going to release this for real now

**Pull request merged**